README.md
#slack-slash

Slack-Slash is a simple framework for handling [slash commands](https://api.slack.com/slash-commands) in [Slack](https://slack.com/). Configure your commands and handlers in one JSON file and the framework does the rest.

## What's a Slash Command?

**/:command [text]**

> Slash commands listen for custom triggers in chat messages across all Slack channels.

> For example, typing **/weather 94070** could send a message to an external URL that would look up the current weather forecast for San Francisco and post it back to the user in Slack.

For more information check out the [Slack API Documentation](https://api.slack.com/slash-commands).

## Adding Handlers

Adding handlers is a simple as requiring a module. In fact, it's even easier.

1. `npm install --save [handler-package-name]`
2. Add a config object for your handler inside `handlers.json`
3. Set your slack token environment variables

## Configuring Handlers

Handlers are defined as objects inside `handlers.json`.

#### Handler Object Properties

```
command:  @{String} slash command typed into slack
pkg:      @{String} module name to require
tokenVar: @{String} environment variable where slack token for slash command is stored
options:  @(Object) any options to pass to the handler [optional]
```

**Example**

```js
// handlers.json
[
  {
    "command": "jira",
    "pkg": "slack-slash-jira",
    "tokenVar": "slack_slash_jira_token",
    "options": {...}
  }
]
```

## Writing Handlers

slack-slash handlers are small modules meant to take the body text from a [Slack Slash Command](https://api.slack.com/slash-commands) and return a message to the Slack user.

### Getting Started

slack-slash handlers export a function. That function takes one or two arguments and must contain a handle method on its prototype that returns the Slack message.

**Sample Handler**
```js
// myCustomHandler.js

module.exports = slashHandler;

var slashHandler = function (token, options) {
  this.token = token;
  this.options = options;
}

slashHandler.prototype.handle = function (req, handleCb) {
  var bodyText = req.body.text;
  handleCb(null, 'Received commmand with text: ' + bodyText);
};

// or if you want to respond with a Slack attachment
slashHandler.prototype.handle = function (req, handleCb) {
  var bodyText = req.body.text;
  handleCb(null, {
    text: 'Received commmand with text: ' + bodyText,
    attachments: [
      {
        'text': 'Handling slash command'
      }
    ]
  });

// or if your response is delayed and should post to response_url instead
slashHandler.prototype.handle = function (req, handleCb) {
  var bodyText = req.body.text;
  handleCb(null, {
    isDelayedResponse: true,
    text: 'Received commmand with text: ' + bodyText,
    attachments: [
      {
        'text': 'Handling slash command'
      }
    ]
  });
};
```

**Handler Arguments**

These arguments are defined in `handlers.json` and get passed into your handler.

- `token` - Token string from configured Slack integration. The token is used to validate requests came from Slack.
- `options` - Optional object with any properties you need to pass to your handler. As of v1.1.0 this object will always contain these two properties in addition to any custom ones that you set:

  - `options.ssFilePath` - The full path where any public files for your handler can be saved `../public/handlerCommand`
  - `options.ssPublicPath` - The relative path which is just the name of your command, `handlerCommand`

### handle(req, callback)

The `handle` method is the entry point to your handler. It will be called by slack-slash with two arguments, the request and a callback function.

**Arguments**

- `req` - This is the request object that contains the Slack post body inside `req.body`.
- `callback(error, message)` - The callback you must call when you are finished handling the request with an error (which can be `null`) and your formatted message.

**Callback Message**

The callback message can be either a string, which can include [formatting supported by Slack](https://api.slack.com/docs/formatting) or a [Slack attachment](https://api.slack.com/docs/attachments).

**Delayed Responses and Multiple Responses**

Slack expects a response from slash commands within 3000 milliseconds. In cases where your handler cannot respond in that timeframe or if you wish to send multiple responses, Slack provides a `response_url` in the request object. In order to let slack-slash know your message should be treated as a delayed response, set the property `isDelayedResponse: true` in the object passed to the callback function. You may send up to 5 delayed responses within 30 minutes of the user's invocation.

### Slack Post Body

This is the data available in the request from Slack.

```js
{
  token: 'xxx',
  team_id: 'xxx',
  team_domain: 'xxx',
  channel_id: 'xxx',
  channel_name: 'xxx',
  user_id: 'xxx',
  user_name: 'xxx',
  command: 'xxx',
  text: 'xxx',
  response_url: 'xxx'
}
```

## Example Handlers

- **[slack-slash-jira](https://github.com/dowjones/slack-slash-jira)** - Get quick info on a Jira ticket right from Slack.

## License

[MIT](/LICENSE.md)


CHANGELOG.md
##Changelog

###1.1.0
- Slack now allows slash commands to respond using extra message formatting or attachments so we now support that. [Learn more.](https://api.slack.com/slash-commands)
- When a handler is registered it the app will now create a public directory with that handler name. The path to this directory gets passed to the handler as an option so now your handler can access its own filesystem.

### 1.0.0
Initial release

routes/index.js
var express  = require('express');
var path     = require('path');
var fs       = require('fs');
var _        = require('lodash');
var request  = require('request');
var nconf    = require('nconf');
var router   = express.Router();
var handlers = require('../handlers');
var handlersObj = {};

var ERRORS = {
  NOT_FOUND: 'Command Not Found',
  UNAUTHORIZED: 'Unauthorized Request'
};

/* CREATE PUBLIC DIRECTORY IF IT DOESN'T EXIST */
var publicPath = path.join(__dirname, '..', 'public');
createPath(publicPath);

module.exports = router;

/* NCONF SETTINGS */
// Parse environment variables
nconf.env();

/* ROUTE DEFINITIONS */

// Slack Slash Commands
router.post('/svc/slack/:command', handleSlackRequest);

// Servo Health Check
router.get('/_health', function (req, res) {
  res.send({
    'status': 'ok'
  });
});

/* REGISTER HANDLERS */
handlers.forEach(registerHandler);

/* FUNCTIONS */

function handleSlackRequest (req, res, next) {
  var command = req.params ? req.params.command : null;
  var slackToken = req.body ? req.body.token : null;
  var err;

  // Verify command exists
  if (!command || !handlersObj[command]) {
    err = new Error(ERRORS.NOT_FOUND);
    err.status = 404;
    return next(err);
  }

  // Verify request came from Slack
  if (!slackToken || slackToken !== handlersObj[command].token) {
    err = new Error(ERRORS.UNAUTHORIZED);
    err.status = 401;
    return next(err);
  }

  // Handle Command
  console.log('Handling command: ' + command);
  handlersObj[command].handle(req, function (handlerError, handlerResponse) {
    respondToSlackRequest(req, res, handlerError, handlerResponse);
  });
}

function respondToSlackRequest(req, res, handlerError, handlerResponse) {

  console.log('handlerResponse:');
  console.log(handlerResponse);

  var defaultResponseObj = {
    response_type: 'ephemeral'
  };
  var resObj;
  var isDelayedResponse;

  // Handle errors and string responses
  if (handlerError || typeof handlerResponse === 'string') {
    resObj = _.assign({text: handlerError ? handlerError : handlerResponse}, defaultResponseObj);
  } else {
    // Handle response objects
    isDelayedResponse = handlerResponse.isDelayedResponse;
    delete handlerResponse.isDelayedResponse;
    resObj = _.assign({}, defaultResponseObj, handlerResponse);    
  }

  if (isDelayedResponse) {
    request({
      url: req.body.response_url,
      method: 'POST',
      body: resObj,
      json: true
    }, function () {
      res.end();
    });
  } else {
    res.json(resObj);
  }
}

function registerHandler (handler) {
  var command = handler.command;
  var pkg = handler.pkg;
  var token = nconf.get(handler.tokenVar);
  var opts = handler.options;
  var dir = path.join(publicPath, command);

  createPath(dir);

  opts.ssFilePath = dir;
  opts.ssPublicPath = command;

  try {
    var SlashCommandHandler = require(pkg);
    handlersObj[command] = new SlashCommandHandler(token, opts);
  }
  catch (e) {
    console.log(e);
  }
}

function createPath(path) {
  try {
    fs.statSync(path);
  } catch (e) {
    fs.mkdirSync(path);
  }
}

server.js
var express      = require('express');
var path         = require('path');
var logger       = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser   = require('body-parser');
var indexHandler = require('./routes/index');
var app = express();

app.set('port', process.env.PORT || 3000);

app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded());
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use('/', indexHandler);

/// catch 404 and forward to error handler
app.use(function(req, res, next) {
  var err = new Error('Not Found');
  err.status = 404;
  next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
  app.use(function(err, req, res) {
    res.status(err.status || 500);
    res.send({
      message: err.message,
      error: err
    });
  });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res) {
  res.status(err.status || 500);
  res.send({
    message: err.message,
    error: {}
  });
});

app.listen(app.get('port'), function() {
  console.log('slack-slash started on port ' + app.get('port'));
});

module.exports = app;